### PR TITLE
direct mode, JS/JSM partition

### PR DESCRIPTION
**Pull request by scottf**
removed enabled check
added getAccountStats api
split JetStream / JetStreamManagement so js cannot be cast to jsm
added direct to subscribe options, changed subscribe to respect direct flag
added ULong
coverage

### CODE REVIEW

File: src/main/java/io/nats/client/api/PurgeResponse.java [hunk 1]
```diff
@@ -42,7 +42,7 @@ public boolean isSuccess() {
      * Returns the number of items purged from the stream
      * @return the count
      */
-    public int getPurgedCount() {
+    public long getPurgedCount() {
```
Comment (scottf) [inline src/main/java/io/nats/client/api/PurgeResponse.java:45]: this is technically an api change

File: src/main/java/io/nats/client/impl/NatsJetStreamMetaData.java [hunk 2]
```diff
@@ -48,12 +48,12 @@ public String toString() {
 
     public NatsJetStreamMetaData(NatsMessage natsMessage) {
         if (!natsMessage.isJetStream()) {
-            throwNotAJetStreamMessage(natsMessage.getReplyTo());
+            throw notAJetStreamMessage(natsMessage.getReplyTo());
```
Comment (scottf) [inline src/main/java/io/nats/client/impl/NatsJetStreamMetaData.java:51]: this is now seen by coverage

File: src/main/java/io/nats/client/impl/NatsJetStreamMetaData.java [hunk 3]
```diff
@@ -48,12 +48,12 @@ public String toString() {
...
         }
 
         String[] parts = natsMessage.getReplyTo().split("\\.");
         if (parts.length < 8 || parts.length > 9 || !"ACK".equals(parts[1])) {
-            throwNotAJetStreamMessage(natsMessage.getReplyTo());
+            throw notAJetStreamMessage(natsMessage.getReplyTo());
```
Comment (scottf) [inline src/main/java/io/nats/client/impl/NatsJetStreamMetaData.java:56]: this is now seen by coverage

File: src/main/java/io/nats/client/impl/NatsJetStreamMetaData.java [hunk 4]
```diff
@@ -137,7 +137,7 @@ public ZonedDateTime timestamp() {
         return timestamp;
     }
 
-    private void throwNotAJetStreamMessage(String subject) {
-        throw new IllegalArgumentException("Message is not a JetStream message.  ReplySubject: <" + subject + ">");
+    private IllegalArgumentException notAJetStreamMessage(String subject) {
+        return new IllegalArgumentException("Message is not a JetStream message.  ReplySubject: <" + subject + ">");
```
Comment (scottf) [inline src/main/java/io/nats/client/impl/NatsJetStreamMetaData.java:141]: done this way for coverage

File: src/examples/java/io/nats/examples/jetstream/NatsJsPubWithOptionsUseCases.java [hunk 5]
```diff
@@ -69,7 +66,7 @@ public static void main(String[] args) {
                     .expectedLastSequence(PublishOptions.UNSET_LAST_SEQUENCE)
                     .messageId(null);
             pa = js.publish(exArgs.subject, "message2".getBytes(), pubOptsBuilder.build());
-            System.out.printf("Published message on subject %s, stream %s, seqno %d.\n",
+            System.out.printf("Published message on subject %s, stream %s, seqno %s.\n",
```
Comment (ColinSullivan1) [inline src/examples/java/io/nats/examples/jetstream/NatsJsPubWithOptionsUseCases.java:69]: %d for seqno

File: src/examples/java/io/nats/examples/jetstream/NatsJsPubWithOptionsUseCases.java [hunk 6]
```diff
@@ -78,15 +75,15 @@ public static void main(String[] args) {
                     .expectedLastSequence(pa.getSeqno()) // last sequence can be found in last ack
                     .messageId("mid3");
             pa = js.publish(exArgs.subject, "message3".getBytes(), pubOptsBuilder.build());
-            System.out.printf("Published message on subject %s, stream %s, seqno %d.\n",
+            System.out.printf("Published message on subject %s, stream %s, seqno %s.\n",
```
Comment (ColinSullivan1) [inline src/examples/java/io/nats/examples/jetstream/NatsJsPubWithOptionsUseCases.java:78]: see above...

File: src/examples/java/io/nats/examples/jetstream/NatsJsPubWithOptionsUseCases.java [hunk 7]
```diff
@@ -78,15 +75,15 @@ public static void main(String[] args) {
...
                     exArgs.subject, pa.getStream(), pa.getSeqno());
 
             // reuse() method clears all the fields, then we set some fields.
             pubOptsBuilder.clearExpected()
                     .expectedLastSequence(pa.getSeqno()) // last sequence can be found in last ack
                     .messageId("mid4");
             pa = js.publish(exArgs.subject, "message4".getBytes(), pubOptsBuilder.build());
-            System.out.printf("Published message on subject %s, stream %s, seqno %d.\n",
+            System.out.printf("Published message on subject %s, stream %s, seqno %s.\n",
```
Comment (ColinSullivan1) [inline src/examples/java/io/nats/examples/jetstream/NatsJsPubWithOptionsUseCases.java:86]: see above

File: src/examples/java/io/nats/examples/jetstream/NatsJsPushSub.java [hunk 8]
```diff
@@ -81,14 +81,11 @@ public static void main(String[] args) {
...
             // Durable means the server will remember where we are if we use that name.
-            PushSubscribeOptions.Builder builder = PushSubscribeOptions.builder();
-            if (exArgs.durable != null) {
-                builder.durable(exArgs.durable);
-            }
-            if (exArgs.deliver != null) {
-                builder.deliverSubject(exArgs.deliver);
-            }
-            PushSubscribeOptions so = builder.build();
+            PushSubscribeOptions so = PushSubscribeOptions.builder()
```
Comment (ColinSullivan1) [inline src/examples/java/io/nats/examples/jetstream/NatsJsPushSub.java:84]: Can you add a comment that null streams and durables are handled?

File: src/main/java/io/nats/client/impl/NatsJetStream.java [hunk 9]
```diff
@@ -412,17 +224,21 @@ NatsJetStreamSubscription createSubscription(String subject, String queueName,
             so = pushSubscribeOptions == null
                     ? PushSubscribeOptions.builder().build()
                     : pushSubscribeOptions;
-            stream = so.getStream(); // might be null, that's ok
+            stream = so.getStream(); // might be null, that's ok (see directBind)
```
Comment (ColinSullivan1) [inline src/main/java/io/nats/client/impl/NatsJetStream.java:227]: bind()

File: src/main/java/io/nats/client/impl/NatsJetStreamMetaData.java (hunk 2 above)
Comment (ColinSullivan1) [reply src/main/java/io/nats/client/impl/NatsJetStreamMetaData.java:51]: maybe add a comment...

File: src/test/java/io/nats/client/SubscribeOptionsTests.java [hunk 10]
```diff
@@ -101,4 +102,25 @@ public void testPullValidation() {
         ConsumerConfiguration cc = ConsumerConfiguration.builder().durable(DURABLE).build();
         PullSubscribeOptions.builder().configuration(cc).build();
     }
+
+    @Test
+    public void testDirectBindCreationErrors() {
```
Comment (ColinSullivan1) [inline src/test/java/io/nats/client/SubscribeOptionsTests.java:107]: Adjust name of test to bind...

File: src/test/java/io/nats/client/impl/JetStreamPubTests.java [hunk 11]
```diff
@@ -80,6 +80,7 @@ public void testPublishVarieties() throws Exception {
 
     private void assertNextMessage(Subscription s, String data) throws InterruptedException {
         Message m = s.nextMessage(DEFAULT_TIMEOUT);
+        System.out.println(m.getClass() + " " + m);
```
Comment (ColinSullivan1) [inline src/test/java/io/nats/client/impl/JetStreamPubTests.java:83]: remove...

**Review by ColinSullivan1 — CHANGES_REQUESTED** [review]
Mostly nitpicks...   Thanks!

File: src/examples/java/io/nats/examples/jetstream/NatsJsPushSub.java (hunk 8 above)
Comment (scottf) [reply src/examples/java/io/nats/examples/jetstream/NatsJsPushSub.java:84]: done

File: src/examples/java/io/nats/examples/jetstream/NatsJsPubWithOptionsUseCases.java (hunk 5 above)
Comment (scottf) [reply src/examples/java/io/nats/examples/jetstream/NatsJsPubWithOptionsUseCases.java:69]: fixed

File: src/examples/java/io/nats/examples/jetstream/NatsJsPubWithOptionsUseCases.java (hunk 6 above)
Comment (scottf) [reply src/examples/java/io/nats/examples/jetstream/NatsJsPubWithOptionsUseCases.java:78]: fixed

File: src/main/java/io/nats/client/impl/NatsJetStreamMetaData.java (hunk 2 above)
Comment (scottf) [reply src/main/java/io/nats/client/impl/NatsJetStreamMetaData.java:51]: fixed

File: src/main/java/io/nats/client/impl/NatsJetStreamMetaData.java (hunk 3 above)
Comment (scottf) [reply src/main/java/io/nats/client/impl/NatsJetStreamMetaData.java:56]: fixed

File: src/main/java/io/nats/client/support/NatsJetStreamConstants.java [hunk 12]
```diff
@@ -48,7 +49,7 @@
     // JSAPI_STREAM_PURGE is the endpoint to purge streams.
     String JSAPI_STREAM_PURGE = "STREAM.PURGE.%s";
 
-    // JSAPI_STREAM_NAMES is the endpoint that will return a list of stream names
+    // JSAPI_STREAM_NAMES.
```
[the remark this answers is not in the document]
Comment (scottf) [reply src/main/java/io/nats/client/support/NatsJetStreamConstants.java:52]: fixed

File: src/main/java/io/nats/client/impl/NatsJetStream.java (hunk 9 above)
Comment (scottf) [reply src/main/java/io/nats/client/impl/NatsJetStream.java:227]: fixed

**Review by ColinSullivan1 — APPROVED** [review]
LGTM.  Thanks!

**Pull request merged**